### PR TITLE
add new tool selenzy v0.2.0

### PR DESCRIPTION
**Pull request by kenza12**
Add new tool **Selenzy v 0.2.0** : Performs enzyme selection from a reaction query : https://github.com/brsynth/selenzy-wrapper

FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

### CODE REVIEW

File: tools/selenzy/selenzy.xml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        #end if
+        --host_taxID '$adv.host_taxID'
+        #if str($adv.enzyme_taxIDs) != "":
+            --enzyme_taxIDs '$adv.enzyme_taxIDs'
+        #end if
+        --nb_ids '$adv.nb_ids'
+        --to_csv '$uniprot_ids_csv'
+    ]]></command>
+    <inputs>
+        <param name="pathway" type="data" format="xml" label="Pathway (SBML)" />
```
Comment (bernt-matthias) [inline tools/selenzy/selenzy.xml:24]: There is also a `sbml` datatype for Galaxy. Does that fit?

File: tools/selenzy/selenzy.xml [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        <section name="adv" title="Advanced Options" expanded="false">
+            <param name="host_taxID" type="integer" value="83333" label="Host taxon ID" />
```
Comment (bernt-matthias) [inline tools/selenzy/selenzy.xml:26]: Please use the `argument` attribute instead of `name` where ever possible.

File: tools/selenzy/selenzy.xml [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            <param name="enzyme_taxIDs" type="text" value="" label="Comma separated taxon IDs of output enzyme sequences" />
+            <param name="nb_ids" type="integer" value="-1" label="Number of uniprot IDs with top score to keep" />
```
Comment (bernt-matthias) [inline tools/selenzy/selenzy.xml:28]: Add `min` and `max` attributes for int and float parameters if possible.

File: tools/selenzy/selenzy.xml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            <param name="nb_targets" type="integer" value="500" label="Number of targets to display in results" />
+            <param name="direction" type="float" value="0" label="Use similiarity values for preferred reaction direction only" />
+            <param name="NoMSA" type="boolean" checked="true" display="checkboxes" label="Do not compute MSA/conservation scores ?" />
```
Comment (bernt-matthias) [inline tools/selenzy/selenzy.xml:31]: You can simplify the command by using `truevalue` and `falsevalue`

File: tools/selenzy/selenzy.xml [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        </section>
+    </inputs>
+    <outputs>
+        <data name="uniprot_ids" format="xml" label="Uniprot IDs" />
```
Comment (bernt-matthias) [inline tools/selenzy/selenzy.xml:35]: This seems also to be sbml datatype, or?

**Comment by bernt-matthias** [comment]
Seems that the tool wants to write in the installation directory:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.8/site-packages/selenzy_wrapper/__main__.py", line 116, in <module>
    entry_point()
  File "/usr/local/lib/python3.8/site-packages/selenzy_wrapper/__main__.py", line 61, in entry_point
    selenzy_pathway(
  File "/usr/local/lib/python3.8/site-packages/selenzy_wrapper/selenzy_wrapper.py", line 56, in selenzy_pathway
    download_and_extract_tar_gz(
  File "/usr/local/lib/python3.8/site-packages/brs_utils/file.py", line 338, in download_and_extract_tar_gz
    extract_tar_gz(filename, dir, member)
  File "/usr/local/lib/python3.8/site-packages/brs_utils/file.py", line 237, in extract_tar_gz
    tar.extractall(dir)
  File "/usr/local/lib/python3.8/tarfile.py", line 2028, in extractall
    self.extract(tarinfo, path, set_attrs=not tarinfo.isdir(),
  File "/usr/local/lib/python3.8/tarfile.py", line 2069, in extract
    self._extract_member(tarinfo, os.path.join(path, tarinfo.name),
  File "/usr/local/lib/python3.8/tarfile.py", line 2143, in _extract_member
    self.makedir(tarinfo, targetpath)
  File "/usr/local/lib/python3.8/tarfile.py", line 2172, in makedir
    os.mkdir(targetpath, 0o700)
PermissionError: [Errno 13] Permission denied: '/usr/local/lib/python3.8/site-packages/selenzy_wrapper/data'
```

(which is impossible here because tools are executed in containers, but should be avoided/forbidden in general).

**Comment by kenza12** [comment]
> Seems that the tool wants to write in the installation directory:
> 
> ```
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.8/runpy.py", line 194, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "/usr/local/lib/python3.8/runpy.py", line 87, in _run_code
>     exec(code, run_globals)
>   File "/usr/local/lib/python3.8/site-packages/selenzy_wrapper/__main__.py", line 116, in <module>
>     entry_point()
>   File "/usr/local/lib/python3.8/site-packages/selenzy_wrapper/__main__.py", line 61, in entry_point
>     selenzy_pathway(
>   File "/usr/local/lib/python3.8/site-packages/selenzy_wrapper/selenzy_wrapper.py", line 56, in selenzy_pathway
>     download_and_extract_tar_gz(
>   File "/usr/local/lib/python3.8/site-packages/brs_utils/file.py", line 338, in download_and_extract_tar_gz
>     extract_tar_gz(filename, dir, member)
>   File "/usr/local/lib/python3.8/site-packages/brs_utils/file.py", line 237, in extract_tar_gz
>     tar.extractall(dir)
>   File "/usr/local/lib/python3.8/tarfile.py", line 2028, in extractall
>     self.extract(tarinfo, path, set_attrs=not tarinfo.isdir(),
>   File "/usr/local/lib/python3.8/tarfile.py", line 2069, in extract
>     self._extract_member(tarinfo, os.path.join(path, tarinfo.name),
>   File "/usr/local/lib/python3.8/tarfile.py", line 2143, in _extract_member
>     self.makedir(tarinfo, targetpath)
>   File "/usr/local/lib/python3.8/tarfile.py", line 2172, in makedir
>     os.mkdir(targetpath, 0o700)
> PermissionError: [Errno 13] Permission denied: '/usr/local/lib/python3.8/site-packages/selenzy_wrapper/data'
> ```
> 
> (which is impossible here because tools are executed in containers, but should be avoided/forbidden in general).

Hi @bernt-matthias , indeed the tool try to download a database locally.

I got the same kind of error with this PR too: https://github.com/galaxyproject/tools-iuc/pull/4314/checks

```
[32m[38;5;15m[1mrptools 5.12.0[0m[38;5;15m (rpextractsink)[0m
[0m
[32m[38;5;15m[1mDownloading cache[0m[0mTraceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/rr_cache/rr_cache.py", line 241, in _check_or_download_cache_to_disk
    raise FileNotFoundError
FileNotFoundError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.9/site-packages/rptools/rpextractsink/__main__.py", line 30, in <module>
    _cli()
  File "/usr/local/lib/python3.9/site-packages/rptools/rpextractsink/__main__.py", line 20, in _cli
    cache = rrCache(['cid_strc'], logger=logger)
  File "/usr/local/lib/python3.9/site-packages/rr_cache/rr_cache.py", line 175, in __init__
    self.load(attrs)
  File "/usr/local/lib/python3.9/site-packages/rr_cache/rr_cache.py", line 194, in load
    rrCache._check_or_download_cache_to_disk(
  File "/usr/local/lib/python3.9/site-packages/rr_cache/rr_cache.py", line 247, in _check_or_download_cache_to_disk
    os_mkdir(cache_dir)
PermissionError: [Errno 13] Permission denied: '/usr/local/lib/python3.9/site-packages/rr_cache/cache'
```
Can I have your opinion/suggestions about this issue please ?

Thanks a lot.

**Comment by kenza12** [comment]
> There are a few options:
> 
>     * if there is a way to configure the download dir then it should be used (some times environment variables are used, e.g. `$TMP` or command line parameters)
> 
>     * if not ask the developers to support this
> 
> 
> Everything else would be a hack ..

Thank you for your answer @bernt-matthias. I contacted the developer and it sounds not a good idea to use a TMP folder for the database because every time that the user will run the command line, the database must be downloaded again. So, I'm trying to figure out another solution and I would like to know why it is forbidden knowing that conda authorizes it ? Do you think there is another alternative to make it work ?

Thanks again.

**Comment by bernt-matthias** [comment]
> Thank you for your answer @bernt-matthias. I contacted the developer and it sounds not a good idea to use a TMP folder for the database because every time that the user will run the command line, the database must be downloaded again. 

Ideally the folder would be configurable with a parameter or env variable (guess `TMP` would be not an ideal choice) such that the DB could also be reused in multi user setups. 

> I would like to know why it is forbidden knowing that conda authorizes it ? Do you think there is another alternative to make it work ?

For containers its just technically impossible to write. 

In my opinion conda does not authorize it, but it just does not forbid it :) .. but it should be forbidden: Just imagine the case where the two users start a tool that writes to a conda env at the same time. Then the content of that file could be pretty messed up.

**Comment by bgruening** [comment]
ping @kenza12

**Comment by kenza12** [comment]
Hi @bernt-matthias , @bgruening , sorry for the delay, I was out of the office. The developer told me that the folder won't be written again if he already exists. And he cannot change that because it's going to take a lot of time and work. Besides, it's not the only tool that needs to be modified (https://github.com/galaxyproject/tools-iuc/pull/4314).
Unfortunately , I will close these 2 pull requests because of this unsolved technical error.

**Comment by bgruening** [comment]
relevant code: https://github.com/galaxyproject/training-material/blob/main/topics/transcriptomics/tutorials/rb-rnaseq/tutorial.md

Can we try to set cache_dir? https://github.com/brsynth/rr_cache/blob/e82cff48d801c11c0d856d2df5b7c875d6f69729/rr_cache/Args.py#L62

**Comment by bgruening** [comment]
Or is it more: https://github.com/brsynth/brs-utils/blob/f28686a250b75a87d393158c78bc76c336aecffc/brs_utils/file.py#L211 but this is using $TEMP as it seems.

**Comment by bgruening** [comment]
Awesome, that are great news!

File: tools/selenzy/selenzy.xml [hunk 6]
```diff
@@ -0,0 +1,97 @@
+<tool id="selenzy-wrapper" name="Selenzyme" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="21.09">
```
Comment (bgruening) [inline tools/selenzy/selenzy.xml:1]:
```suggestion
<tool id="selenzy" name="Selenzyme" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="21.09">
```

Comment (bgruening) [reply tools/selenzy/selenzy.xml:1]: Is it selenzy or Selenzyme?

Comment (kenza12) [reply tools/selenzy/selenzy.xml]: Hi, the tool ID is selenzy-wrapper actually and the tool name is Selenzyme. I just changed the directory name to selenzy-wrapper.

**Comment by nsoranzo** [comment]
From the lint CI job:

```
WARNING: Repository name [selenzy-wrapper] invalid. Repository names must contain only lower-case letters, numbers and underscore.
```

Can you please replace selenzy-wrapper with selenzy_wrapper in directory/file names and repository/tool ids?

@bernt-matthias Should this an error instead of a warning?

File: tools/selenzy-wrapper/.shed.yml [hunk 7]
```diff
@@ -6,6 +6,6 @@ long_description: 'Selenzy performs enzyme sequence selection from a reaction qu
   The tool can be queried using a reaction template such as the reaction rules in
   RetroRules(https://github.com/brsynth/RRParser). This feature makes this tool especially
   useful in combination with RetroPath2.0 (https://github.com/brsynth/RetroPath2-wrapper).'
-name: selenzy
+name: selenzy-wrapper
```
Comment (bgruening) [inline tools/selenzy-wrapper/.shed.yml:9]: I personally don't like to have the name `wrapper` in the ID or the repo name, but if you like it we can proceed and merge. Imho its a redundant suffix that admins need to deal with when configuring the tools etc...

**Comment by bernt-matthias** [comment]
> From the lint CI job:
> 
> ```
> WARNING: Repository name [selenzy-wrapper] invalid. Repository names must contain only lower-case letters, numbers and underscore.
> ```
> 
> Can you please replace selenzy-wrapper with selenzy_wrapper in directory/file names and repository/tool ids?
> 
> @bernt-matthias Should this an error instead of a warning?

It  says "must" .. so probably yes.

File: tools/selenzy-wrapper/.shed.yml (hunk 7 above)
Comment (kenza12) [reply tools/selenzy-wrapper/.shed.yml]: Yes I understand but this tool was wrapped from another program, so we prefer to keep this ID. Thank you.

**Review by bgruening — APPROVED** [review]

**Comment by bgruening** [comment]
Thanks a lot @kenza12 for addressing all our small and big comments!

**Pull request merged**